src/pip/_internal/req/req_install.py

```diff
@@ -457,6 +459,7 @@ def is_wheel_from_cache(self) -> bool:
     # Things valid for sdists
     @property
     def unpacked_source_directory(self) -> str:
+        assert self.source_dir, f"No source dir for {self}"
```

Comment:
I guess this is one place where we could raise an AssertionError where the existing code would not fail: if `self.source_dir` was an empty string. But `assert self.source_dir` appears many other times in this file, so I think is okay.